### ISSUE TITLE
OGRFeatureQuery::Evaluate error evaluating ORed conditions where the last is true

### ISSUE THREAD

**Issue by elpaso** [OSGeo/gdal#3919]
From https://github.com/qgis/QGIS/issues/43181

Reproduce:

```python
def test_ogr_shape_filter_expression():

    ds = ogr.GetDriverByName(
        'ESRI Shapefile').CreateDataSource('/vsimem/test.shp')
    lyr = ds.CreateLayer('test', geom_type=ogr.wkbPoint)
    field = ogr.FieldDefn('fclass', ogr.OFTString)
    lyr.CreateField(field)
    g = ogr.Geometry(ogr.wkbPoint)
    g.AddPoint(0, 1)
    f = ogr.Feature(lyr.GetLayerDefn())
    f.SetGeometry(g)
    f.SetField('fclass', 'x')
    assert lyr.CreateFeature(f) == ogr.OGRERR_NONE
    lyr.SetAttributeFilter(
        """(fclass = 'a' OR fclass = 'b') OR (fclass = 'c' OR fclass = 'd' OR fclass = 'x')""")

    assert lyr.GetFeatureCount() == 1

    ds = None
    ogr.GetDriverByName('ESRI Shapefile').DeleteDataSource('/vsimem/test.shp')

```

Note that this works:
```python
lyr.SetAttributeFilter(
        """(fclass = 'a' OR fclass = 'b') OR (fclass = 'x' OR fclass = 'd' OR fclass = 'c')""")
```

**Comment by elpaso** [comment]
@rouault I've tracked down the issue to `swq_expr_node *swq_expr_node::Evaluate`, the `poRetNode = poOp->pfnEvaluator( this, &(apoValues[0]) );` is called on the first two nodes even when the `nSubExprCount` is 3. If that's an issue in the expression parser or in the evaluation code I can't tell.

**Comment by rouault** [comment]
> is called on the first two nodes even when the `nSubExprCount` is 3

normally nSubExprCount should be 2. It might be >= 3 in a temporary way, and the swq_fixup_expression() should address that to reconstruct a binary tree. Which failed here. See fix in #3921
